delfin/drivers/ibm/ds8k/rest_handler.py

```diff
@@ -0,0 +1,123 @@
+from delfin import exception
+from delfin.drivers.utils.rest_client import RestClient
+
+LOG = logging.getLogger(__name__)
+
+
+class RestHandler(RestClient):
+    REST_TOKEN_URL = '/api/v1/tokens'
+
+    def __init__(self, **kwargs):
+        self.session_lock = threading.Lock()
+        super(RestHandler, self).__init__(**kwargs)
+
+    def call_with_token(self, url, data, method):
+        auth_key = None
+        if self.session:
+            auth_key = self.session.headers.get('X-Auth-Token', None)
+            if auth_key:
+                self.session.headers['X-Auth-Token'] \
+                    = cryptor.decode(auth_key)
+        res = self.do_call(url, data, method)
+        if auth_key:
+            self.session.headers['X-Auth-Token'] = auth_key
+        return res
+
+    def call(self, url, data=None, method=None):
+        try:
+            res = self.call_with_token(url, data, method)
+            if res.status_code == 401:
+                LOG.error("Failed to get token,status_code:%s,error_mesg:%s" %
+                          (res.status_code, res.text))
+                self.login()
+                res = self.call_with_token(url, data, method)
+            elif res.status_code == 503:
+                raise exception.InvalidResults(res.text)
+            return res
+        except Exception as e:
+            LOG.error("Method:%s,url:%s failed: %s" % (method, url,
+                                                       six.text_type(e)))
+            raise e
+
+    def get_rest_info(self, url, data=None, method='GET'):
+        result_json = None
+        res = self.call(url, data, method)
+        if res.status_code == 200:
+            result_json = res.json()
+        return result_json
+
+    def login(self):
+        try:
+            data = {"username": self.rest_username,
+                    "password": cryptor.decode(self.rest_password)
+                    }
+            data = {'request': {'params': data}}
+            with self.session_lock:
+                if self.session is None:
+                    self.init_http_head()
+                res = self.call_with_token(
+                    RestHandler.REST_TOKEN_URL, data, 'POST')
+                if res.status_code == 200:
+                    result = res.json()
+                    self.session.headers['X-Auth-Token'] = \
+                        cryptor.encode(result.get('token').get('token'))
+                else:
+                    LOG.error("Login error. URL: %(url)s，Reason: %(reason)s.",
+                              {"url": RestHandler.REST_TOKEN_URL,
+                               "reason": res.text})
+                    if 'Authentication has failed' in res.text:
+                        raise exception.InvalidUsernameOrPassword()
+                    else:
+                        raise exception.StorageBackendException(res.text)
+        except Exception as e:
+            LOG.error("Login error: %s", six.text_type(e))
+            raise e
```

Comment:
Since there is password plaintext in the local variable `data`, for security requirement, suggest to add `finally: data = None` at the end of `try` block. And please help to add `finally: data = None; kwargs = None` at the end of `try` block in function `delfin/drivers/utils/rest_client.py: do_call()` too.

Reply:
done

Reply:
done